### ISSUE TITLE
Login throttling using JdbcAuditTrailManager won't work

### ISSUE THREAD

**Issue by mateofacu** [Jasig/cas#1347]
#### Context
Project: Migration from CAS 3.5.2 to CAS 4.1.2
I am applying the same customizations I have done  to v4.1.2.
One of those is Throttling functionality which is working on the older version.

#### Problem
I notice that the login throttling mecanism in this version won't work with JdbcAuditTrailManager . 

In the docs. JdbcAuditTrailManager and  InspektrThrottledSubmissionByIpAddressAndUsernameHandlerInterceptorAdapter is a valid combination. 

#### Analisys 
I have found two problems: 

_1- Username problem_:
When JdbcAuditTrailManager  is active, the AUD_USER field in the database get it value in the format "`<username>`+password" for UserPasswordCredentials.

InspektrThrottledSubmissionByIpAddressAndUsernameHandlerInterceptorAdapter query the username field  in the following format "[username: `<username>`]". 

Because of this, the query will always return 0 .

_2- IP Adress problem_: 

In class InspektrThrottledSubmissionByIpAddressAndUsernameHandlerInterceptorAdapter the database is queried using this snippet:
```java
final List<Timestamp> failures = this.jdbcTemplate.query(
                    sqlQueryAudit,
                    new Object[]{request.getRemoteAddr(), userToUse, this.authenticationFailureCode,
                            this.applicationCode, cutoff.getTime()},
                    new int[]{Types.VARCHAR, Types.VARCHAR, Types.VARCHAR, Types.VARCHAR, Types.TIMESTAMP},
                    new RowMapper<Timestamp>() {
                        @Override
                        public Timestamp mapRow(final ResultSet resultSet, final int i) throws SQLException {
                            return resultSet.getTimestamp(1);
                        }
                    });
```

 the AUD_CLIENT_IP is compared against  `request.getRemoteAddr()` .

The framework Inspektr uses `request.getRemoteAddr()` by default  to save audit information. But it allows users to change the request header from where this value should be taken.  

Based on this approach, we have configured inspektr to use a different request header to get the user ip address by modifying ClientInfoThreadLocalFilter in web.xml: 

```xml
<filter-class>com.github.inspektr.common.web.ClientInfoThreadLocalFilter</filter-class>
  	<init-param>
  		<param-name>alternativeIpAddressHeader</param-name>
  		<param-value>X-Forwarded-For</param-value>
  	</init-param>
...
```
So now, inspektr  will take X-Forwarded-For header to get the user ip address.

But  InspektrThrottledSubmissionByIpAddressAndUsernameHandlerInterceptorAdapter will still get the ip address using `request.getRemoteAddr()`. 

So basically , this class should get the client ip address in the same way that "recordThrottle" method does, using org.jasig.cas.web.support.ClientInfo object.

```java
 final ClientInfo clientInfo = ClientInfoHolder.getClientInfo();
...
clientInfo.getClientIpAddress(),
```

**Comment by mmoayyed** [comment]
1 is already fixed. Try with 4.1.3 SNAPSHOT. 

I am not sure I understand 2. Inspektr should populate remote user with the value from the header. Does it not do that?

**Comment by mateofacu** [comment]
Good to know about  1 ! Thanks.

I updated the description on number too trying to clarify it. I apologize for my english. 
If it is not clear enough , I will make another try.

**Comment by mmoayyed** [comment]
No, you did perfectly :) Thanks. Reading the description, this is certainly sounding like an issue. Would you want to submit a patch/PR? If not, I could probably handle this sometime this week, but of course, it would be more promising if it came from you since you already, assuming, have the environment to setup and test it. 

If you are unsure of how to submit PRs, let me know and I'll guide you. Here is what you need basically:
http://jasig.github.io/cas/developer/Contributor-Guidelines.html

Always submit a PR against master. We'll cherry-pick the change back into 4.1.3. 

We are planning to release 4.1.3 in about a week. So if you could get this to us by then, that would be super.

**Comment by mateofacu** [comment]
Thanks misagh!  I didn't have the setup for gradle environment on Master branch. 
And I was having some problems to set it up on eclipse. 

Anyway, I will finish the setup in case I can contribute to the proyect.

**Comment by mmoayyed** [comment]
No worries. You should be able to leverage the fix with 4.1.3 SNAPSHOT until it's officially released. 

Appreciate future contributions. To set things up with Intellij, simply execute `gradle idea` at the root, and then open the project in IDEA. Should work just fine.

**Comment by mateofacu** [comment]
Good to know. "gradle eclipse" has some bugs related to .properties  config files. 
I will try with idea. 
Thanks!

**Comment by mateofacu** [comment]
Misagh I have verified new release 4.1.3.
Issue 1 (addressed  here)  is still present.
In master branch I can see there is a fix for this but It seems it wasn't included in this release.

**Comment by mmoayyed** [comment]
My mistake, you are perfectly right. Looks like the fix was only applied to master. Would you be willing to submit a PR for 4.1 to handle this problem?

**Comment by mmoayyed** [comment]
I believe if you applied the changes in this file:
https://github.com/Jasig/cas/blob/a90504e741c4325f43fe8f567f5f889e978f00bd/cas-server-core-authentication/src/main/java/org/jasig/cas/authentication/UsernamePasswordCredential.java

to the 4.1.x branch, the problem would go away. It's a very simple change if you want to take it up yourself.

**Comment by mateofacu** [comment]
UserNamePasswordCredential and this files I think should do the trick
https://github.com/Jasig/cas/blob/a90504e741c4325f43fe8f567f5f889e978f00bd/cas-server-webapp-throttle/src/main/java/org/jasig/cas/web/support/InspektrThrottledSubmissionByIpAddressAndUsernameHandlerInterceptorAdapter.java

Is not a problem to make the fix. 

If there is no hurry on this, I will finish the CAS dev environment setup following the http://jasig.github.io/cas/developer/Contributor-Guidelines.html for Branch 4.1.X and will submit the PR.

**Comment by mmoayyed** [comment]
There is no rush. We do patch releases once every 30 days :) Technically, you have about a month, and once you submit, you would be table to test the 4.1.4-SNAPSHOT in about 40 minutes from the time your PR merges in. So take your time. Thanks!

**Comment by mmoayyed** [comment]
Happy holidays. Do we have an update on this?

**Comment by mateofacu** [comment]
Happy holidays !  Not yet, too much work on this end of the year. 
But don't worry, I will do this on the next days. I think we are still on time.

**Comment by mateofacu** [comment]
Misagh I have implemented the fix on my cas fork.
I will have access to a real balanced testing environment on Monday 4 (where this solution is needed) . Then I will submit the PR.

If you are in a hurry to release version 4.1.4 I can create a maven shadow project  to test this implementation before.

**Comment by mmoayyed** [comment]
No you're fine. We have about 2 weeks before the next patch release. Carry on :)
